rsocket-routing-http-bridge/src/main/java/io/rsocket/routing/http/bridge/core/AbstractHttpRSocketFunction.java

```diff
@@ -0,0 +1,141 @@
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+import io.rsocket.routing.client.spring.RoutingRSocketRequester;
+import io.rsocket.routing.client.spring.RoutingRSocketRequesterBuilder;
+import io.rsocket.routing.common.spring.ClientTransportFactory;
+import io.rsocket.routing.common.spring.TransportProperties;
+import io.rsocket.routing.http.bridge.config.RSocketHttpBridgeProperties;
+import io.rsocket.transport.ClientTransport;
+import org.apache.commons.logging.Log;
+
+import org.springframework.beans.factory.ObjectProvider;
+
+import static io.rsocket.routing.http.bridge.core.AbstractHttpRSocketFunction.Transport.WEBSOCKET;
+import static org.apache.commons.logging.LogFactory.getLog;
+
+/**
+ * Base class for HTTP-RSocket bridge functions.
+ *
+ * @author Olga Maciaszek-Sharma
+ * @since 0.3.0
+ */
+abstract class AbstractHttpRSocketFunction<I, O> implements Function<I, O> {
+
+	public static final String TRANSPORT_KEY = "transport";
+	public static final String PORT_KEY = "port";
+	public static final String HOST_KEY = "host";
+	public static final String MAPPING_PATH_KEY = "mapping-path";
+
+	protected final Log LOG = getLog(getClass());
+
+	protected RoutingRSocketRequester defaultRequester;
+	protected RoutingRSocketRequesterBuilder requesterBuilder;
+	ObjectProvider<ClientTransportFactory> transportFactories;
+	protected final RSocketHttpBridgeProperties properties;
+	protected final Duration timeout;
+
+	protected AbstractHttpRSocketFunction(RoutingRSocketRequesterBuilder requesterBuilder,
+			RoutingRSocketRequester defaultRequester, ObjectProvider<ClientTransportFactory> transportFactories,
+			RSocketHttpBridgeProperties properties) {
+		this.requesterBuilder = requesterBuilder;
+		this.properties = properties;
+		timeout = properties.getTimeout();
+		this.transportFactories = transportFactories;
+		this.defaultRequester = defaultRequester;
+	}
+
+	protected void logTimeout(String address, String route) {
+		if (LOG.isDebugEnabled()) {
+			LOG.debug(String
+					.format("Timeout occurred while retrieving RSocket response from address: %s, route: %s. Response was not retrieved within %s", address, route, timeout));
+		}
+	}
+
+	protected void logException(Throwable error, String address, String route) {
+		if (LOG.isErrorEnabled())
+			LOG.error(String
+					.format("Exception occurred while retrieving RSocket response from address: %s, route: %s", address, route), error);
+	}
+
+	protected RoutingRSocketRequester getRequester(String brokerHeader) {
+		if (brokerHeader != null) {
+			TransportProperties broker = buildBroker(brokerHeader);
+			ClientTransport clientTransport = transportFactories
+					.orderedStream()
+					.filter(factory -> factory.supports(broker))
+					.findFirst()
+					.map(factory -> factory.create(broker))
+					.orElseThrow(() -> new IllegalStateException("Unknown transport: " + broker));
+			RoutingRSocketRequester requesterForBroker = requesterBuilder
```

Comment:
Building a requester for every request is fine for now, but this should likely be cached for some period of time based on the header.

Reply:
Ok. I'll create an issue and get back to it after checking the GW integration.

Reply:
https://github.com/rsocket-routing/rsocket-routing-broker/issues/25